tests/unit/lms/services/jwt_test.py

```diff
@@ -0,0 +1,80 @@
+import copy
+import datetime
+
+import jwt
+import pytest
+from freezegun import freeze_time
+from pytest import param
+
+from lms.services.exceptions import ExpiredJWTError, InvalidJWTError
+from lms.services.jwt import JWTService
+
+
+class TestJWTService:
+    def test_it_can_decode_a_jwt_from_encoded_jwt(self, svc):
+        original_payload = {"key": "value"}
+        jwt_str = svc.encode_with_secret(
+            original_payload, "test_secret", datetime.timedelta(hours=1)
+        )
+
+        decoded_payload = svc.decode_with_secret(jwt_str, "test_secret")
+
+        assert decoded_payload == original_payload
+
+    @pytest.mark.parametrize(
+        "error,kwargs",
+        (
+            param(InvalidJWTError, {"secret": "WRONG"}, id="wrong_secret"),
+            param(InvalidJWTError, {"algorithm": "HS384"}, id="wrong_algorithm"),
+            param(
+                ExpiredJWTError,
+                {"lifetime": datetime.timedelta(hours=-1)},
+                id="already_expired",
+            ),
+        ),
+    )
+    def test_decode_raises_for_bad_jwts(self, svc, error, kwargs):
+        jwt_params = {"lifetime": datetime.timedelta(hours=24)}
+        jwt_params.update(**kwargs)
+
+        jwt_str = self.encode_jwt({"key": "value"}, **jwt_params)
+
+        with pytest.raises(error):
+            svc.decode_with_secret(jwt_str, "test_secret")
+
+    def encode_jwt(
+        self,
+        payload,
+        lifetime,
+        secret="test_secret",
+        algorithm="HS256",
+    ):
+        """Return payload encoded to a jwt with secret and algorithm."""
+        payload = copy.deepcopy(payload)
+
+        payload["exp"] = datetime.datetime.utcnow() + lifetime
+
+        return jwt.encode(payload, secret, algorithm=algorithm)
+
+    @freeze_time("2022-04-04")
+    def test_encode_with_secret(self, svc, jwt):
```

Comment:
Here I replaced all of the encoding jwt test with this one:

- `test_it_can_decode_a_jwt_from_encoded_jwt` already tests the encode <-> decode full flow.
- Here only asserting how we use the `jwt` library.
- Using this `freeze_time` library. Maybe a very small usage to warrant a new dependency?
 
I found that when time is involved what freeze gun does is what I want to do.

Reply:
I think we should give this a shot, and maybe separately look for tests which could be simplified by using this. I can't think of any off the top of my head, but I know we've ended up with all sorts of nonsense with fixtures for `now` and patching out `datetime` and things like that.